### ISSUE TITLE
prefer-const makes destructuring redundant

### ISSUE THREAD

**Issue by kaiye** [airbnb/javascript#1958]
The rule `prefer-const` with the option  `{"destructuring": "any"}` will report an error on `unit is never reassigned.  Use 'const' instead`

```js
function foo(value) {
  let [, v, unit] = value.toLowerCase().match(/(\d+)([a-z]*)/) || [];
  if (unit === 'rpx') {
    v *= 1.5;
  }
  return v ? parseInt(v, 10) : 0;
}
```
If I modify the code for fixing this error, the code will become not elegant.
```js
function foo2(value) {
  const result = value.toLowerCase().match(/(\d+)([a-z]*)/) || [];
  let [, v] = result;
  const [, , unit] = result;
  if (unit === 'rpx') {
    v *= 1.5;
  }
  return v ? parseInt(v, 10) : 0;
}
```
Is there any better suggestion?

**Comment by ljharb** [comment]
```js
function foo(value) {
  const [, v, unit] = value.toLowerCase().match(/(\d+)([a-z]*)/) || [];
  const parsed = parseInt(v, 10);
  if (!parsed) {
    return 0;
  }
  return unit === 'rpx' ? parsed * 1.5 : parsed;
}
```

Note how I’ve also fixed the bug of multiplying by 1.5 on an unparsed string, as well as a string that is truthy but parses to NaN will now return 0 :-)

**Comment by petersendidit** [comment]
Can also avoid the interim variables and assign directly to the object properties. 
```js
function foo2(obj) {
  const { a, b, c, d, e, f } = obj;
  return {
    ...obj,
    d: bar(a, b, c, d),
    e: bar(a, b, c, e),
    f: bar(a, b, c, f),
  };
}
```
